Add Workflow Registry Metrics and remove executions running counter

patrickhuie19 opened the pull request:
The executions running gauge can be misleading since it is updated on heartbeat, which is async to workflow deletes through the workflow registry. 

I saw we're lacking some metrics in the workflow registry and so added coverage.

patrickhuie19 committed:
changed engineCleanup from best effort based on service health to definitive

patrickhuie19 committed:
removing total executions running metric

patrickhuie19 commented on core/services/workflows/engine.go:
```diff
@@ -1153,7 +1153,6 @@ func (e *Engine) heartbeat(ctx context.Context) {
 			return
 		case <-ticker.C:
 			e.metrics.incrementEngineHeartbeatCounter(ctx)
-			e.metrics.updateTotalWorkflowsGauge(ctx, e.stepUpdatesChMap.len())
```
I am removing this metric entirely. Read PR description for why

cedric-cordenier replied:
Hmmm... this still feels like it could be useful @patrickhuie19 -- I can imagine scenarios where the number of executions is just increasing, and not going down such that we'll eventually exhaust the queue. In that case this metric would help us diagnose what's going on.

You mentioned that this relates to deletes of the workflow engine -- can you say more about how? I'm not following entirely

cedric-cordenier replied:
Note that one source of confusion here is that this name of this metric (total workflows running) doesn't correspond to what it's measuring (number of executions running an in instance of the workflow engine); maybe that's the source of confusion here?

patrickhuie19 replied:
It seems to be more confusing than helpful. The name is part of it. The metric is updated on a heartbeat, which is async to engine.Start() and engine.Close(). 

>I can imagine scenarios where the number of executions is just increasing, and not going down such that we'll eventually exhaust the queue

Thanks for the example use case. Would we be able to satisfy this by looking at the completed duration histogram metrics and observing no executions are finishing?

patrickhuie19 commented on core/services/workflows/engine.go:
```diff
@@ -1266,7 +1265,7 @@ const (
 	defaultQueueSize            = 100000
 	defaultNewWorkerTimeout     = 2 * time.Second
 	defaultMaxExecutionDuration = 10 * time.Minute
-	defaultHeartbeatCadence     = 5 * time.Minute
```
5 minutes is anemic

patrickhuie19 commented on core/services/workflows/syncer/handler.go:
```diff
@@ -752,20 +752,18 @@ func (h *eventHandler) forceUpdateSecretsEvent(
-// tryEngineCleanup attempts to stop the workflow engine for the given workflow ID.  Does nothing if the
-// workflow engine is not running.
-func (h *eventHandler) tryEngineCleanup(wfID string) error {
-	if h.engineRegistry.IsRunning(wfID) {
```
This had logic based on whether or not the underlying `Ready` implementation in the embedded services.StateMachine would return an error. The Engine has no subservices, Close() isn't marked as syncronous with Ready in services.StateMachine, and if the Engine wasn't Ready, we would do nothing anyways.

If we want to stay flexible to checking the status of Ready in the future, we could add some type of sync loop to check status and delete when Ready?

cedric-cordenier replied:
@patrickhuie19 Are we subtly changing behaviour here? We'll return an error now if we can't retrieve the engine from the registry; whereas previous we would have ignored that

patrickhuie19 replied:
Yes. I'll change it so that engineCleanup errors under the same condition as previously to keep the scope of this PR smaller. The tradeoff with handling a lacking engine instance this gracefully is that it might fly under our radar when engine instances are getting cleared from the registry before they should.

patrickhuie19 replied:
e75d11740ae4fe08f6a9f83751c2c9a54565ce19

patrickhuie19 committed:
removing test in syncer relying on engine.Ready before Close

patrickhuie19 committed:
Adding metrics to workflow registry syncer and lacking test coverage

bolekk reviewed (commented):
Maybe I'm too tired but I'm really struggling to make sense of the description of this PR.

What I see is going on:
1. You are removing "workflowsRunningGauge" because it is pretty useless. It's counting current number of executions at a given point in time, updated only with the heartbeat, correct?
2. You are adding a bunch of new counters to track all events in the Workflow Registry. Those make sense to me, especially "totalWorkflowsGauge". It won't cover workflows added outside the Registry but I guess that's fine.

What I don't understand is how this is going to help debug why metrics are lingering after workflow deletion. You say that other metrics hang around too, not only the executions gauge. Are you able to reproduce it locally? What new signal are you expecting to get once we start running this in prod?

patrickhuie19 committed:
fixing test

patrickhuie19 commented:
> Maybe I'm too tired but I'm really struggling to make sense of the description of this PR.
> 
> What I see is going on:
> 
>     1. You are removing "workflowsRunningGauge" because it is pretty useless. It's counting current number of executions at a given point in time, updated only with the heartbeat, correct?
> 
>     2. You are adding a bunch of new counters to track all events in the Workflow Registry. Those make sense to me, especially "totalWorkflowsGauge". It won't cover workflows added outside the Registry but I guess that's fine.
> 
> 
> What I don't understand is how this is going to help debug why metrics are lingering after workflow deletion. You say that other metrics hang around too, not only the executions gauge. Are you able to reproduce it locally? What new signal are you expecting to get once we start running this in prod?

Thanks, simplified the description. Your points 1 and 2 are correct. There were multiple levers into the changes in this PR, and they are worthwhile on their own outside of the context of the time-series persisting post deletion bug we're tracking down.

bolekk reviewed (dismissed):
Thanks!

cedric-cordenier commented on core/services/workflows/syncer/handler.go:
```diff
@@ -346,11 +360,21 @@ func (h *eventHandler) Handle(ctx context.Context, event Event) error {
+		metrics.incrementRegisterCounter(ctx)
+
+		// intentionally without workflow specific labels
+		h.metrics.updateTotalWorkflowsGauge(ctx, int64(h.engineRegistry.Size()))
```
Yes but slightly misleading: this really only measures workflows that have been started by the syncer -- we should add a increment in the workflows/delegate to give an accurate picture

patrickhuie19 replied:
e75d11740ae4fe08f6a9f83751c2c9a54565ce19

cedric-cordenier commented on core/services/workflows/syncer/handler.go:
```diff
@@ -427,11 +474,20 @@ func (h *eventHandler) Handle(ctx context.Context, event Event) error {
 
+		metrics.incrementDeleteCounter(ctx)
+		// intentionally without workflow specific labels
+		h.metrics.updateTotalWorkflowsGauge(ctx, int64(h.engineRegistry.Size()))
```
Nice :)

cedric-cordenier commented on core/services/workflows/syncer/handler.go:
```diff
@@ -406,11 +445,19 @@ func (h *eventHandler) Handle(ctx context.Context, event Event) error {
 			return err
 		}
 
+		metrics.incrementActivateCounter(ctx)
```
@patrickhuie19 Shouldn't we increment the total workflows counter here and decrement when we pause?

patrickhuie19 replied:
No, I don't think so. The total workflows gauge is to track the number of created/deleted workflows from the perspective of the syncer.

patrickhuie19 replied:
Ah, no, I was wrong. It wasn't obvious to me that activate and pause delete the associated engine from the registry. Why do we have it implemented like that?

patrickhuie19 replied:
e75d11740ae4fe08f6a9f83751c2c9a54565ce19

krehermann commented on core/services/workflows/syncer/handler.go:
```diff
@@ -216,13 +217,20 @@ func NewEventHandler(
+
+	m, err := initMonitoringResources()
+	if err != nil {
+		lggr.Fatalw("Failed to initialize monitoring resources", "err", err)
```
hmm. this will kill the node. that seems unreasonable, we don't know what else is running. 

where is this init all use to live?

krehermann replied:
if it really should block the whole node, then panic

if not, then critical

patrickhuie19 replied:
For this particular error to get invoked it signals a potentially catastrophic regression in Beholder impl. But agreed we don't know what else is running

patrickhuie19 replied:
e75d11740ae4fe08f6a9f83751c2c9a54565ce19

patrickhuie19 committed:
separating engine from workflows metrics. wiring an engineRegistry around to make that possible

patrickhuie19 committed:
fixing import cycle. workflow -/-> syncer

patrickhuie19 committed:
fixing web test

patrickhuie19 committed:
fixing other tests in unexpected place

patrickhuie19 committed:
missing initialization of metrics struct in handler

patrickhuie19 committed:
lint

krehermann commented on core/services/workflows/delegate.go:
```diff
@@ -38,11 +40,20 @@ func (d *Delegate) AfterJobCreated(jb job.Job) {}
 func (d *Delegate) BeforeJobDeleted(spec job.Job) {}
 
-func (d *Delegate) OnDeleteJob(context.Context, job.Job) error { return nil }
+func (d *Delegate) OnDeleteJob(ctx context.Context, spec job.Job) error {
```
it seems like we are making a conscious choice to always return nil.

i can't understand why that would be the case from the surrounding code.  please add a comment for future readers and maintainers

patrickhuie19 replied:
Unsure, I opted to stay backwards compatible. @cedric-cordenier for sanity check as I think you implemented this initially

cedric-cordenier replied:
I don't think we need to stay backwards compatible -- anything upstream should be handling errors appropriately

krehermann commented on core/services/workflows/delegate.go:
```diff
@@ -80,6 +91,12 @@ func (d *Delegate) ServicesForSpec(ctx context.Context, spec job.Job) ([]job.Ser
+
+	err = d.engineRegistry.Add(spec.WorkflowSpec.WorkflowID, engine)
+	if err != nil {
+		d.logger.Errorf("delegate failed to register workflow engine for workflow name: %s id: %s: %v", cfg.WorkflowName.String(), cfg.WorkflowID, err)
```
? why aren't we returning the error? it if failed to add can we really continue?

patrickhuie19 replied:
This approach is backwards compatible - if we failed to add it to the engineRegistry the previous functionality is still locked in.

cedric-cordenier replied:
We should return an error here IMO -- this should really never happen and if it does this op should fail

krehermann commented on core/services/workflows/syncer/handler.go:
```diff
@@ -346,11 +369,21 @@ func (h *eventHandler) Handle(ctx context.Context, event Event) error {
 
+		metrics.incrementRegisterCounter(ctx)
+
+		// intentionally without workflow specific labels
```
nit/ finish the sentence. i don't know why we would have that intention

patrickhuie19 replied:
Sentence is complete - we have that intention because it is tracking how many concurrent workflows are registered and viable for execution via a trigger. It's not specific to any workflow.

cedric-cordenier replied:
Probably Keith is pointing out that the more valuable comment here is /why/ we don't want workflow specific labels

cedric-cordenier commented on core/services/workflows/syncer/monitoring.go:
```diff
@@ -0,0 +1,107 @@
+func initMonitoringResources() (l *workflowRegistryMetricsLabeler, err error) {
+	m := &workflowRegistryMetrics{}
+
+	m.activateCounter, err = beholder.GetMeter().Int64Counter("platform_workflow_syncer_register")
```
platform_workflow_syncer_register -> platform_workflow_syncer_activate?

cedric-cordenier commented on core/services/workflows/syncer/monitoring.go:
```diff
@@ -0,0 +1,107 @@
+	updateCounter             metric.Int64Counter
+}
+
+func initMonitoringResources() (l *workflowRegistryMetricsLabeler, err error) {
```
Is it worth adopting the following pattern for these metrics:

platform_workflow_syncer_events_register
platform_workflow_syncer_events_update
etc?

patrickhuie19 replied:
Yes - if we want to instrument measurements outside the scope of the events handler. thanks

patrickhuie19 committed:
engine registry failures now exposed to consumers

patrickhuie19 committed:
adding comments

patrickhuie19 committed:
Merge branch 'develop' into fix/execution-num-metric

patrickhuie19 committed:
lint

patrickhuie19 committed:
fixed test after lint

patrickhuie19 committed:
Merge branch 'develop' into fix/execution-num-metric
